hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java

```diff
@@ -0,0 +1,412 @@
+                : rewardRate / (totalStakedRewardStart / HBARS_TO_TINYBARS);
+        log.info(
+                "The reward rate for the period was {} tb ({} tb/hbar for nodes with in-range stake, given {} total stake reward start)",
+                rewardRate,
+                perHbarRate,
+                totalStakedRewardStart);
+
+        // Calculate the updated stake and reward sum history for each node
+        long newTotalStakedStart = 0L;
+        long newTotalStakedRewardStart = 0L;
+        long maxStakeOfAllNodes = 0L;
+        final Map<Long, StakingNodeInfo> updatedNodeInfos = new HashMap<>();
+        for (final var nodeNum : nodeIds.stream().sorted().toList()) {
+            var currStakingInfo = stakingInfoStore.getForModify(nodeNum);
+
+            // The return value is the reward rate (tinybars-per-hbar-staked-to-reward) that will be paid to all
+            // accounts who had staked-to-reward for this node long enough to be eligible in the just-finished period
+            final var newRewardSumHistory = calculateRewardSumHistory(
+                    currStakingInfo,
+                    perHbarRate,
+                    stakingConfig.maxDailyStakeRewardThPerH(),
+                    stakingConfig.requireMinStakeToReward());
+            final var newPendingRewardRate = newRewardSumHistory.pendingRewardRate();
+            currStakingInfo = currStakingInfo
+                    .copyBuilder()
+                    .rewardSumHistory(newRewardSumHistory.rewardSumHistory())
+                    .build();
+            log.info(
+                    "   > Non-zero reward sum history is now {}",
+                    () -> readableNonZeroHistory(newRewardSumHistory.rewardSumHistory()));
+
+            final var oldStakeRewardStart = currStakingInfo.stakeRewardStart();
+            final var pendingRewardHbars =
+                    (oldStakeRewardStart - currStakingInfo.unclaimedStakeRewardStart()) / HBARS_TO_TINYBARS;
+            final var recomputedStake =
+                    com.hedera.node.app.service.token.impl.staking.PeriodStakingUtils.computeNextStake(currStakingInfo);
+            currStakingInfo = currStakingInfo
+                    .copyBuilder()
+                    .stake(recomputedStake.stake())
+                    .stakeRewardStart(recomputedStake.stakeRewardStart())
+                    .build();
+            final var newStakeRewardStart = recomputedStake.stakeRewardStart();
+            final var nodePendingRewards = pendingRewardHbars * newPendingRewardRate;
+            log.info(
+                    "For node{}, the tb/hbar reward rate was {} for {} pending, with stake reward start {} -> {}",
+                    nodeNum,
+                    newPendingRewardRate,
+                    nodePendingRewards,
+                    oldStakeRewardStart,
+                    newStakeRewardStart);
+            stakeInfoHelper.increaseUnclaimedStakeRewards(nodeNum, nodePendingRewards, stakingInfoStore);
```

Comment:
I think these are pending rewards for next period and not `unclaimedRewards` . It seems the logic here is different

Reply:
Good catch, I used the wrong object for this! Will fix

Reply:
Done